File: android/src/main/kotlin/com/appdynamics/appdynamics_mobilesdk/features/InfoPoints.kt
/*
 * Copyright (c) 2021. AppDynamics LLC and its affiliates.
 * All rights reserved.
 *
 */

package com.appdynamics.appdynamics_mobilesdk.features

import androidx.annotation.NonNull
import com.appdynamics.appdynamics_mobilesdk.AppDynamicsMobileSdkPlugin
import com.appdynamics.eumagent.runtime.Instrumentation
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

fun AppDynamicsMobileSdkPlugin.beginCall(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val callId = properties["callId"] as String
    val className = properties["className"] as? String
    val methodName = properties["methodName"] as? String
    val methodArgs = (properties["methodArgs"] as? ArrayList<*>) ?: emptyList<Any>()

    if (className == null) {
        result.error(
            "500",
            "Agent trackCall() failed.",
            "Please provide a valid class name."
        )
        return
    }

    if (methodName == null) {
        result.error(
            "500",
            "Agent trackCall() failed.",
            "Please provide a valid method name."
        )
        return
    }

    val tracker =
        Instrumentation.beginCall(false, className, methodName, *methodArgs.toTypedArray())
    callTrackers[callId] = tracker

    result.success(null)
}


fun AppDynamicsMobileSdkPlugin.endCallWithSuccess(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val callId = properties["callId"] as String
    val value = properties["result"]

    val tracker = callTrackers[callId]
    tracker!!.reportCallEndedWithReturnValue(value);
    callTrackers.remove(callId)

    result.success(null)
}

fun AppDynamicsMobileSdkPlugin.endCallWithError(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val callId = properties["callId"] as String
    val error = properties["error"] as HashMap<*, *>

    val dict = mutableMapOf<String, Any>("is_error" to true)
    val message = error["message"] as? String
    val stackTrace = error["stackTrace"] as? String


    if (message != null) {
        dict["message"] = message
    }

    if (stackTrace != null) {
        dict["stackTrace"] = stackTrace
    }

    val tracker = callTrackers[callId]
    tracker!!.reportCallEndedWithReturnValue(JSONObject(dict.toMap()).toString())
    callTrackers.remove(callId)

    result.success(null)
}

File: ios/Classes/Features/InfoPoints.swift
import ADEUMInstrumentation
import Flutter

extension SwiftAppDynamicsAgentPlugin {
  func beginCall(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let callId = properties["callId"] as! String
    
    guard let className = properties["className"] as? String else {
      let error = FlutterError(code: "500", message: "Agent trackCall() failed.", details: "Please provide a valid class name.")
      result(error)
      return
    }
    
    guard let methodName = properties["methodName"] as? String else {
      let error = FlutterError(code: "500", message: "Agent trackCall() failed.", details: "Please provide a valid methodName value.")
      result(error)
      return
    }
    
    let methodArgs: [Any] = (properties["methodArgs"] as? [Any]) ?? []
    
    let tracker = ADEumInstrumentation.beginCall(className, methodName: methodName, withArguments: methodArgs)
    callTrackers[callId] = tracker
    
    result(nil)
  }
  
  func endCallWithSuccess(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    
    
    let callId = properties["callId"] as! String
    let value = properties["result"]!
    
    ADEumInstrumentation.endCall(callTrackers[callId], withValue: value)
    callTrackers[callId] = nil
    
    result(nil)
  }
  
  func endCallWithError(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    
    
    let callId = properties["callId"] as! String
    let error = properties["error"] as! [String: Any]
    
    var dict: [String: Any] = ["is_error": true]
    
    if let message = error["message"] as? String {
      dict["message"] = message
    }
    
    if let stackTrace = error["stackTrace"] as? String {
      dict["stackTrace"] = stackTrace
    }
    
    let jsonData = try! JSONSerialization.data(withJSONObject: dict, options: .prettyPrinted)
    let jsonString = String(data: jsonData, encoding: .utf8)
    
    // We will not be using the native `endCall` because it requires a native iOS error.
    // Instead, we end the call with a custom error-like payload.
    ADEumInstrumentation.endCall(callTrackers[callId], withValue: jsonString)
    callTrackers[callId] = nil;
    
    result(nil)
  }
}
